hwy/ops/generic_ops-inl.h

```diff
@@ -1249,6 +1250,43 @@ HWY_API V PopulationCount(V v) {
 }
 #endif
 
+HWY_API Vec128<int64_t,1> operator*(Vec128<int64_t,1> x, Vec128<int64_t,1> y) {
+    return Set(Simd<int64_t,1,0>(), GetLane(x)*GetLane(y));
+}
+HWY_API Vec128<uint64_t,1> operator*(Vec128<uint64_t,1> x, Vec128<uint64_t,1> y) {
+    return Set(Simd<uint64_t,1,0>(), GetLane(x)*GetLane(y));
+}
+
+#ifndef HWY_NATIVE_I64MULLO
+
+
+HWY_API Vec128<uint64_t> operator*(Vec128<uint64_t> x, Vec128<uint64_t> y) {
+  Vec128<uint32_t> x32 = BitCast(Simd<uint32_t,4,0>(), x);
+  Vec128<uint32_t> y32 = BitCast(Simd<uint32_t,4,0>(), y);
+  Vec128<uint64_t> lolo = MulEven(x32, y32);
+  Vec128<uint32_t> lohi = BitCast(Simd<uint32_t,4,0>(), MulEven(x32, BitCast(Simd<uint32_t,4,0>(), ShiftRight<32>(y))));
+  Vec128<uint32_t> hilo = BitCast(Simd<uint32_t,4,0>(), MulEven(BitCast(Simd<uint32_t,4,0>(), ShiftRight<32>(x)), y32));
+  return lolo + ShiftLeft<32>(BitCast(Simd<uint64_t,2,0>(), lohi + hilo));
+}
+HWY_API Vec128<int64_t> operator*(Vec128<int64_t> x, Vec128<int64_t> y) {
+  return BitCast(Simd<int64_t,2,0>(), BitCast(Simd<uint64_t,2,0>(), x) * BitCast(Simd<uint64_t,2,0>(), y));
+}
+#if HWY_TARGET == HWY_AVX2
+
+HWY_API Vec256<uint64_t> operator*(Vec256<uint64_t> x, Vec256<uint64_t> y) {
+  Vec256<uint32_t> x32 = BitCast(Simd<uint32_t,8,0>(), x);
+  Vec256<uint32_t> y32 = BitCast(Simd<uint32_t,8,0>(), y);
+  Vec256<uint64_t> lolo = MulEven(x32, y32);
+  Vec256<uint32_t> lohi = BitCast(Simd<uint32_t,8,0>(), MulEven(x32, BitCast(Simd<uint32_t,8,0>(), ShiftRight<32>(y))));
+  Vec256<uint32_t> hilo = BitCast(Simd<uint32_t,8,0>(), MulEven(BitCast(Simd<uint32_t,8,0>(), ShiftRight<32>(x)), y32));
+  return lolo + ShiftLeft<32>(BitCast(Simd<uint64_t,4,0>(), lohi + hilo));
```

Comment:
The casts back to `uint32_t` before adding and casting again to `uint64_t` and shifting by 32 may look funny, because we'd get the exact same answer out of the exact same number of instructions by avoiding the casts on x86_64 by dropping the casts and doing the add on `uint64_t` directly.

I did this because I figured in theory doing 2X 32 bit adds should be cheaper than doing X 64 bit adds. While latencies and throughputs are the same, perhaps it uses slightly less power.
Also, on Haswell, `vpaddq` requires the complex decoder, while `vpaddd` does not. Thus in unrolled code (not operating out of the uop cache), `vpaddd` gets better throughput:
https://uops.info/html-instr/VPADDD_YMM_YMM_YMM.html
https://uops.info/html-instr/VPADDQ_YMM_YMM_YMM.html

Reply:
Interesting. I'd think the uop cache is the common case. To better predict performance, we can enable the llvm-mca tool in godbolt to predict throughput of an actual code block, not just one instruction.
[Example](https://gcc.godbolt.org/z/n-KcQ-)

I suspect that a 64-bit add is OK here.

Reply:
Note that llvm-mca does not model the front end, thus if the difference is expected to mostly be in the front end, we wouldn't expect llvm-mca to help.

I tried with a simple example, but I'd probably have to try a specific degenerate example to get integer adds to actually become the bottleneck.